File: src/wings/tools/builtin/write.py
"""Create or overwrite a file."""

from pathlib import Path

from pydantic import BaseModel, Field

from wings.tools.base import ToolContext
from wings.tools.decorator import tool


class WriteInput(BaseModel):
    file_path: str = Field(description="Absolute path to the file to write")
    content: str = Field(description="Content to write to the file")


@tool(
    name="write",
    description="Create or overwrite a file with the given content",
    search_hint="write /path/to/file",
    destructive=True,
)
async def write_file(input: WriteInput, context: ToolContext) -> str:
    path = Path(input.file_path)
    if not path.is_absolute():
        path = Path(context.working_dir) / path

    existed = path.exists()
    old_text = path.read_text() if existed else ""
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(input.content)

    action = "Updated" if existed else "Created"
    lines = input.content.split("\n")
    line_count = len(lines)

    if not existed:
        # Show all lines as added
        result = [f"{action} {path} ({line_count} line(s))"]
        for line in lines[:30]:
            result.append(f"         +{line.rstrip()}")
        if len(lines) > 30:
            result.append(f"         ... ({len(lines) - 30} more lines)")
        return "\n".join(result)

    # For updates, show summary + changed line count
    old_lines = old_text.split("\n")
    added = max(0, line_count - len(old_lines))
    removed = max(0, len(old_lines) - line_count)
    parts = []
    if added:
        parts.append(f"+{added} line(s)")
    if removed:
        parts.append(f"-{removed} line(s)")
    change_desc = ", ".join(parts) if parts else "content replaced"

    return f"{action} {path}: {change_desc} ({line_count} lines)"

File: src/tools/builtin/write.ts
/** Create or overwrite a file. */

import { existsSync, mkdirSync, readFileSync, writeFileSync, statSync } from "node:fs";
import { isAbsolute, join, dirname } from "node:path";

import { z } from "zod";

import { buildTool } from "../types.ts";

interface WriteInput {
  file_path: string;
  content: string;
}

function linesChanged(old: string, new_: string): string {
  const oldLines = old.split("\n");
  const newLines = new_.split("\n");
  const added = Math.max(0, newLines.length - oldLines.length);
  const removed = Math.max(0, oldLines.length - newLines.length);

  const parts: string[] = [];
  if (added) parts.push(`Added ${added} line${added !== 1 ? "s" : ""}`);
  if (removed) {
    const caps = !added ? "Removed" : "removed";
    parts.push(`${caps} ${removed} line${removed !== 1 ? "s" : ""}`);
  }
  if (parts.length === 0) parts.push("Content replaced");
  return parts.join(", ");
}

export const writeTool = buildTool({
  name: "write",
  description: "Create or overwrite a file with the given content",
  search_hint: "write /path/to/file",
  is_destructive: true,
  inputSchema: z.object({
    file_path: z.string().describe("Absolute path to the file to write"),
    content: z.string().describe("Content to write to the file"),
  }),
  async call(input: WriteInput, context) {
    let pathStr = input.file_path;
    if (!isAbsolute(pathStr)) pathStr = join(context.working_dir, pathStr);

    const existed = existsSync(pathStr);

    // Stale detection for existing files
    if (existed) {
      const cachedMtime = context.read_cache[pathStr];
      if (cachedMtime === undefined) {
        return `Error: must read ${pathStr} before writing to it. Use the read tool first.`;
      }
      const currentMtime = statSync(pathStr).mtimeMs / 1000;
      if (currentMtime > cachedMtime) {
        return `Error: ${pathStr} was modified since last read (cached: ${cachedMtime}, current: ${currentMtime}). Re-read the file and try again.`;
      }
    }

    const oldText = existed ? readFileSync(pathStr, "utf-8") : "";
    mkdirSync(dirname(pathStr), { recursive: true });
    writeFileSync(pathStr, input.content);

    // Update read cache so the file is "fresh" after write
    context.read_cache[pathStr] = statSync(pathStr).mtimeMs / 1000;

    const newLines = input.content.split("\n");
    const lineCount = newLines.length;

    if (!existed) {
      // New file — show first lines as preview with + prefix
      const result = [`Wrote ${lineCount} line${lineCount !== 1 ? "s" : ""} to ${pathStr}`];
      const preview = Math.min(lineCount, 10);
      for (let i = 0; i < preview; i++) {
        result.push(`         +${newLines[i]!.trimEnd()}`);
      }
      if (lineCount > preview) {
        result.push(`    \u2026 +${lineCount - preview} lines`);
      }
      return result.join("\n");
    }

    // Update — show summary + diff-like preview
    const summary = linesChanged(oldText, input.content);
    const result = [summary, `(in ${pathStr})`];

    // Show a diff-like view: new lines with + where different
    const oldLines = oldText.split("\n");
    const maxShow = Math.min(Math.max(oldLines.length, lineCount), 30);
    let changed = 0;
    for (let i = 0; i < maxShow; i++) {
      const old = i < oldLines.length ? oldLines[i]!.trimEnd() : "";
      const newL = i < lineCount ? newLines[i]!.trimEnd() : "";
      if (old !== newL) {
        if (old) result.push(`    ${String(i + 1).padStart(6)} -${old}`);
        if (newL) result.push(`    ${String(i + 1).padStart(6)} +${newL}`);
        changed++;
      } else if (changed > 0) {
        result.push(`    ${String(i + 1).padStart(6)}  ${old}`);
      }
    }
    if (Math.max(lineCount, oldLines.length) > maxShow) {
      result.push(`    \u2026 (${Math.max(lineCount, oldLines.length) - maxShow} more lines)`);
    }

    return result.join("\n");
  },
});
